
syntax = "proto3";

package reddit;

message User {
    // a human readable id
    string id = 1;
}

message Post {
    // a human readable id
    string id = 1;
    
    // title, some text, one of video or image url
    string title = 2;
    string content = 3;
    oneof media {
        string image_url = 4;
        string video_url = 5;
    }

    // optional author
    optional string user_id = 6;

    // a score, can be negative
    int32 score = 7;

    // a state: normal, locked, hidden
    enum PostState {
        NORMAL = 0;
        LOCKED = 1;
        HIDDEN = 2;
    }
    PostState state = 8;

    // a publication date
    string publication_date = 9;
}

message Comment {
    // a human readable id
    string id = 1;

    // the post or comment it is under
    oneof parent {
        string parent_post_id = 2;
        string parent_comment_id = 3;
    }

    // must have an author
    string user_id = 4;

    // must have some text
    string content = 5;

    // have a score
    int32 score = 6;

    // can be normal or hidden
    enum CommentState {
        NORMAL = 0;
        HIDDEN = 1;
    }
    CommentState state = 7;

    // a publication date
    string publication_date = 8;
}

message GetTopCommentsRequest {
    // the post to get the top comments for
    string id = 1;
}

message RetrieveTopCommentsRequest {
    int32 id = 1;
    int32 limit = 2;
}

message RetrieveTopCommentsResponse {
    repeated Comment comments = 1;
}


service RedditService {
    rpc CreatePost(Post) returns (Post);
    rpc UpvotePost(Post) returns (Post);
    rpc DownvotePost(Post) returns (Post);
    rpc RetrievePostContent(Post) returns (Post);
    rpc CreateComment(Comment) returns (Comment);
    rpc UpvoteComment(Comment) returns (Comment);
    rpc DownvoteComment(Comment) returns (Comment);
    rpc RetrieveTopComments(RetrieveTopCommentsRequest) returns (RetrieveTopCommentsResponse);
}







